File: tests/js/vertexai/test_reference.ts
/**
 * Conformance test: Reference instrumentation for Vertex AI.
 */
import { trace } from "@opentelemetry/api";
import { flushAndShutdownOtel, setupOtel } from "../otel";

const nodeProcess = globalThis.process as NodeJS.Process & {
  env: Record<string, string | undefined>;
  exit(code?: number): never;
};

const MOCK_LLM_URL = nodeProcess.env.MOCK_LLM_URL!;
const tracer = trace.getTracer("gen_ai.reference");

// Patch fetch before any SDK imports to intercept auth and redirect HTTPS→HTTP.
const _originalFetch = globalThis.fetch;
globalThis.fetch = async function (input: any, init?: any) {
  let url =
    typeof input === "string"
      ? input
      : input instanceof URL
        ? input.href
        : input.url;

  // Mock OAuth token endpoint so the SDK gets a fake access token
  if (url.includes("oauth2.googleapis.com") || url.includes("/token")) {
    return new Response(
      JSON.stringify({
        access_token: "mock-token",
        token_type: "Bearer",
        expires_in: 3600,
      }),
      { status: 200, headers: { "content-type": "application/json" } }
    );
  }

  // Redirect HTTPS to HTTP for mock server
  url = url.replace(/^https:\/\//, "http://");
  return _originalFetch(url, init);
};

async function main() {
  console.log("=== Reference: Vertex AI JS Conformance Test ===");
  const otel = setupOtel();

  const { VertexAI } = await import("@google-cloud/vertexai");
  const { createFakeGoogleAuth } = await import(
    "@google-cloud/vertexai/build/src/testing/fake_google_auth"
  );
  const mockUrl = new URL(MOCK_LLM_URL);
  const vertexai = new VertexAI({
    project: "test-project",
    location: "us-central1",
    apiEndpoint: `${mockUrl.hostname}:${mockUrl.port}`,
  });
  (vertexai as any).googleAuth = createFakeGoogleAuth({
    accessToken: "mock-token",
  });
  const requestModel = "gemini-2.0-flash";
  const model = vertexai.getGenerativeModel({ model: requestModel });

  // Scenario: chat
  console.log("  [chat] basic content generation (reference)");
  await tracer.startActiveSpan("chat gemini-2.0-flash", async (span) => {
    span.setAttribute("gen_ai.operation.name", "chat");
    span.setAttribute("gen_ai.provider.name", "vertex_ai");
    span.setAttribute("gen_ai.request.model", requestModel);
    const result = await model.generateContent("Say hello.");
    const resp = result.response;
    const responseModel = (resp as any).modelVersion ?? requestModel;
    span.setAttribute("gen_ai.response.model", responseModel);
    const candidate = resp.candidates?.[0];
    if (candidate?.finishReason) {
      span.setAttribute("gen_ai.response.finish_reasons", [candidate.finishReason]);
    }
    if (resp.usageMetadata) {
      if (resp.usageMetadata.promptTokenCount) {
        span.setAttribute("gen_ai.usage.input_tokens", resp.usageMetadata.promptTokenCount);
      }
      if (resp.usageMetadata.candidatesTokenCount) {
        span.setAttribute("gen_ai.usage.output_tokens", resp.usageMetadata.candidatesTokenCount);
      }
    }
    const text = candidate?.content?.parts?.[0]?.text ?? "";
    console.log(`    -> ${text.slice(0, 60)}`);
    span.end();
  });

  // Scenario: streaming chat
  console.log("  [chat_streaming] streaming content generation (reference)");
  await tracer.startActiveSpan("chat gemini-2.0-flash", async (span) => {
    span.setAttribute("gen_ai.operation.name", "chat");
    span.setAttribute("gen_ai.provider.name", "vertex_ai");
    span.setAttribute("gen_ai.request.model", requestModel);
    const result = await model.generateContentStream("Tell me a joke.");
    let text = "";
    for await (const chunk of result.stream) {
      const part = chunk.candidates?.[0]?.content?.parts?.[0]?.text;
      if (part) text += part;
    }
    const aggregated = await result.response;
    const responseModel = (aggregated as any).modelVersion ?? requestModel;
    span.setAttribute("gen_ai.response.model", responseModel);
    const candidate = aggregated.candidates?.[0];
    if (candidate?.finishReason) {
      span.setAttribute("gen_ai.response.finish_reasons", [candidate.finishReason]);
    }
    if (aggregated.usageMetadata) {
      if (aggregated.usageMetadata.promptTokenCount) {
        span.setAttribute("gen_ai.usage.input_tokens", aggregated.usageMetadata.promptTokenCount);
      }
      if (aggregated.usageMetadata.candidatesTokenCount) {
        span.setAttribute("gen_ai.usage.output_tokens", aggregated.usageMetadata.candidatesTokenCount);
      }
    }
    console.log(`    -> ${text.slice(0, 60)}`);
    span.end();
  });

  await flushAndShutdownOtel(otel);
}

main().catch((e) => { console.error(e); nodeProcess.exit(1); });

File: tests/python/vertexai/test_reference.py
"""Conformance test: reference instrumentation for Vertex AI.

Exercises: chat, chat_streaming
against a mock Vertex AI server, with manual OTel spans.
"""

import os
import warnings

from otel_setup import setup_otel, flush_and_shutdown
from opentelemetry import trace

MOCK_BASE_URL = os.environ["MOCK_LLM_URL"]

# The Vertex AI gapic REST transport defaults to HTTPS. Monkey-patch the
# transport to use plain HTTP so we can talk to the local mock server.
from google.cloud.aiplatform_v1.services.prediction_service.transports.rest import (  # noqa: E402
    PredictionServiceRestTransport,
)

_original_rest_init = PredictionServiceRestTransport.__init__


def _patched_rest_init(self, **kwargs):
    kwargs.setdefault("url_scheme", "http")
    return _original_rest_init(self, **kwargs)


PredictionServiceRestTransport.__init__ = _patched_rest_init

_reference_tracer = trace.get_tracer("gen_ai.reference")


def _mock_host():
    """Return host:port from MOCK_BASE_URL (strip scheme)."""
    return MOCK_BASE_URL.replace("http://", "").replace("https://", "")


def _init_vertexai():
    """Initialize Vertex AI SDK pointing at the mock server."""
    import vertexai
    from google.auth.credentials import AnonymousCredentials

    vertexai.init(
        project="test-project",
        location="us-central1",
        credentials=AnonymousCredentials(),
        api_endpoint=_mock_host(),
        api_transport="rest",
    )


def run_chat():
    """Scenario: basic chat completion with reference instrumentation."""
    from vertexai.generative_models import GenerativeModel

    print("  [chat] basic chat completion via Vertex AI (reference)")
    request_model = "gemini-2.0-flash"
    with _reference_tracer.start_as_current_span("chat gemini-2.0-flash") as span:
        span.set_attribute("gen_ai.operation.name", "chat")
        span.set_attribute("gen_ai.provider.name", "vertex_ai")
        span.set_attribute("gen_ai.request.model", request_model)
        with warnings.catch_warnings():
            warnings.simplefilter("ignore", DeprecationWarning)
            warnings.simplefilter("ignore", UserWarning)
            model = GenerativeModel(request_model)
            response = model.generate_content("Say hello.")
        response_model = response.to_dict().get("modelVersion")
        if response_model:
            span.set_attribute("gen_ai.response.model", response_model)
        else:
            span.set_attribute("gen_ai.response.model", request_model)
        if response.candidates and response.candidates[0].finish_reason:
            span.set_attribute("gen_ai.response.finish_reasons", [str(response.candidates[0].finish_reason.name)])
        if hasattr(response, "usage_metadata") and response.usage_metadata:
            if response.usage_metadata.prompt_token_count:
                span.set_attribute("gen_ai.usage.input_tokens", response.usage_metadata.prompt_token_count)
            if response.usage_metadata.candidates_token_count:
                span.set_attribute("gen_ai.usage.output_tokens", response.usage_metadata.candidates_token_count)
        print(f"    -> {response.text[:60]}")


def run_chat_streaming():
    """Scenario: streaming chat completion with reference instrumentation."""
    from vertexai.generative_models import GenerativeModel

    print("  [chat_streaming] streaming chat completion via Vertex AI (reference)")
    request_model = "gemini-2.0-flash"
    with _reference_tracer.start_as_current_span("chat gemini-2.0-flash") as span:
        span.set_attribute("gen_ai.operation.name", "chat")
        span.set_attribute("gen_ai.provider.name", "vertex_ai")
        span.set_attribute("gen_ai.request.model", request_model)
        with warnings.catch_warnings():
            warnings.simplefilter("ignore", DeprecationWarning)
            warnings.simplefilter("ignore", UserWarning)
            model = GenerativeModel(request_model)
            text = ""
            last_chunk = None
            for chunk in model.generate_content("Tell me a joke.", stream=True):
                text += chunk.text
                last_chunk = chunk
        if last_chunk:
            response_model = last_chunk.to_dict().get("modelVersion")
            if response_model:
                span.set_attribute("gen_ai.response.model", response_model)
            else:
                span.set_attribute("gen_ai.response.model", request_model)
        if last_chunk and last_chunk.candidates and last_chunk.candidates[0].finish_reason:
            span.set_attribute("gen_ai.response.finish_reasons", [str(last_chunk.candidates[0].finish_reason.name)])
        if last_chunk and hasattr(last_chunk, "usage_metadata") and last_chunk.usage_metadata:
            if last_chunk.usage_metadata.prompt_token_count:
                span.set_attribute("gen_ai.usage.input_tokens", last_chunk.usage_metadata.prompt_token_count)
            if last_chunk.usage_metadata.candidates_token_count:
                span.set_attribute("gen_ai.usage.output_tokens", last_chunk.usage_metadata.candidates_token_count)
        print(f"    -> {text[:60]}")


def main():
    print("=== Reference: Vertex AI Conformance Test ===")

    tp, lp, mp = setup_otel()
    # NO instrument() call – reference instrumentation only
    _init_vertexai()

    run_chat()
    run_chat_streaming()

    flush_and_shutdown(tp, lp, mp)


if __name__ == "__main__":
    main()
